Update fullnode documentation to include pruning and archive fallback (#13063)
MystenLabs/sui: doc/src/build/fullnode.md

    ```

Your Full node starts on: `http://127.0.0.1:9000`.

# Pruning 

## Object Pruning

Object pruning involves the removal of older object versions from the fullnode db. 
New object versions are added to the db during transaction execution (making older versions ready for garbage collection) and can become a 
problem if not pruned, as it can slow down the performance of the db and take up too much disk space. 

> Object pruning works in the background by tailing checkpoint stream, identifying each
> checkpoint which contains object versions which are now eligible for pruning.

Reviewer: what does "tailing checkpoint stream" mean? 
```suggestion
Sui identifies the objects that are eligible for pruning in each checkpoint, and then performs the pruning in the background.
```

Author: i meant to day - it works by continuously reading checkpoints in the background but you captured it much better in words

> A sui node can enable object pruning by adding this `authority-store-pruning-config` config:

Reviewer: ```suggestion
You can enable pruning for a Sui node by adding the `authority-store-pruning-config` config:
```
Please specify the name and location of the file to add it to

Author: added `fullnode.yaml`

```yaml
authority-store-pruning-config:
  # Number of epoch dbs to keep 
  # Not relevant for object pruning
  num-latest-epoch-dbs-to-retain: 3

>   # How often the epoch db pruning task should run

Reviewer: The amount of time, in seconds, between running the object pruning task.

  # Not relevant for object pruning
  epoch-db-pruning-period-secs: 3600

>   # Number of latest epochs before which object pruning is allowed to run.

Reviewer: latest=previous?
```suggestion
  # Number of epochs to wait before performing object pruning.
```

>   # When this is 0, pruner is allowed to prune old object versions as soon
>   # as possible. This is what we also call aggressive pruning and is the most effective
>   # garbage collection method which results in the samllest disk usage possible. 
>   # This is recommended setting for validator nodes as older object versions aren't
>   # needed for transaction execution.

Reviewer: ```suggestion
  # When set to 0, Sui prunes old object versions as soon
  # as possible. This is also called *aggressive pruning*, and results in the most effective
  # garbage collection method with lowest disk usage possible. 
  # This is the recommended setting for Sui Validator nodes since older object versions aren't
  # necessary to execute transactions.
```

>   # When this is 1, pruner will only prune object versions from transaction checkpoints
>   # older than the current epoch. In general, when this is N (where N >= 1), pruner 
>   # will only prune object versions from checkpoints upto current - N epoch. 
>   # This means it is possible to have multiple versions of an object to be present 
>   # in the db. It is a recommended setting for fullnodes as they may have to serve 
>   # rpc requests which require looking up objects by id and version (instead of just latest
>   # version). However, if a fullnode is not serving such rpc requests it should also 
>   # use aggressive pruning

Reviewer: ```suggestion
  # When set to1, Sui prunes only object versions from transaction checkpoints
  # previous to the current epoch. In general, when set to N (where N >= 1), Sui prunes  
  # only object versions from checkpoints up to `current - N` epoch. 
  # It is therefore possible to have multiple versions of an object present 
  # in the database. This setting is recommended for Sui Full nodes as they might need to serve 
  # RPC requests that require looking up objects by ID and Version (rather than just latest
  # version). However, if your Full node does not serve RPC requests you should then also enable  
  # aggressive pruning.
```

  num-epochs-to-retain: 0

>   # Advanced setting: Maximum number of checkpoints in one batch of pruning run. This is

Reviewer: ```suggestion
  # Advanced setting: Maximum number of checkpoints to prune in a batch. The default
```

>   # mostly not needed to be changed and default is good for most use cases

Reviewer: ```suggestion
  # settings are appropriate for most for most use cases.
```
When would one need to change the setting?

  max-checkpoints-in-batch: 10
  # Advanced setting: Maximum number of transactions in one batch of pruning run. This is

>   # mostly not needed to be changed and default is good for most use cases

Author: remaining transactions get picked up in the next iteration of the pruning task. This task continuously runs in the background. The above limit is for maximum number of checkpoints to prune in one iteration

  max-transactions-in-batch: 1000
```

> ## Transaction Pruning

Reviewer: ```suggestion
## Transaction pruning
```

> Transaction pruning is another type of pruning that removes old transactions and effects from the db. 

Reviewer: ```suggestion
Transaction pruning removes previous transactions and effects from the database. 
```

> Checkpoints which are created periodically contain transaction and their effects. 
> Transaction pruning works in the background by tailing checkpoint stream as well.
> Full nodes and validators can enable transaction pruning by adding can enable transaction pruning
> by adding `num_epochs_to_retain_for_checkpoints` to their `authority-store-pruning-config` config:

Reviewer: ```suggestion
Sui periodically creates checkpoints. Each checkpoint contains the transactions that occurred during the checkpoint and their associated effects. 
Sui performs transaction pruning in the background after checkpoints complete. 
You can enable transaction pruning for your Full node or Validator node by adding  `num_epochs_to_retain_for_checkpoints` to the `authority-store-pruning-config` config for the node:
```

```yaml
authority-store-pruning-config:
  num-latest-epoch-dbs-to-retain: 3
  epoch-db-pruning-period-secs: 3600
  num-epochs-to-retain: 0
  max-checkpoints-in-batch: 10
  max-transactions-in-batch: 1000
  # Number of latest epochs before which transaction and effects could be pruned.
  # When this is N (where N >= 2), pruner can prune transaction and effects from 
  # checkpoints upto current - N epoch. Transaction and effects from current and
  # previous epoch are never pruned. N = 2 is a recommended setting for validators
  # and fullnodes not serving rpc requests.
  num_epochs_to_retain_for_checkpoints: 2

>   # Ensures that individual SST files periodically go through the compaction process.

Reviewer: will everyone know what an SST file is?

Author: used database files isntead

  # This helps reclaim disk space and avoid fragmentation issues
  periodic-compaction-threshold-days: 1
```

## Archival Fallback

Once various full nodes start pruning historical transactions and effects, it may not be always possible for
their peers to catch up transaction and effects via state synchronization. And for this a node needs to fallback
downloading this data from a remote aws s3 bucket. This archive has all transaction and effects data since 
genesis and lags behind the latest network checkpoint by ten minutes. Enabling this is very highly recommended

> for all nodes. A sui node can enable this automatic archive fallback by adding the following config:

Reviewer: After Sui starts performing transaction pruning on Full nodes to remove historical transactions and their effects, it might not be possible for peer nodes to catch up with the transactions and effects via synchronization. Instead, peer nodes can fall back by downloading and restoring a [database snapshot](../build/snapshot.md). 
The snapshot is an archive of all transaction data on Sui, trailing behind the latest checkpoint by 10 minutes. You should enable this for all nodes as a best practice. To configure a Sui node to automatically fall back to a snapshot image, add the following to your config file.

What is the name and location of this file, the fullnode.yaml?

Author: I don't want to use the word snapshot here because that is a slightly different concept. But used all the other parts of your suggested change

```yaml
state-archive-read-config:
  - object-store-config:
      object-store: "S3"
      # Use mysten-testnet-archives for testnet 
      # Use mysten-mainnet-archives for mainnet
      bucket: "mysten-<testnet|mainnet>-archives"
      # Use your aws account access key id
      aws-access-key-id: "<AWS_ACCESS_KEY_ID>"

>       # Use your aws account secret access key

Reviewer: AWS

      aws-secret-access-key: "<AWS_SECRET_ACCESS_KEY>"
      aws-region: "us-west-2"
      object-store-connection-limit: 20
    # How many objects to read ahead when catching up  
    concurrency: 5
    # Whether to prune local state based on latest checkpoint in archive.
    # This should stay false for most use cases
    use-for-pruning-watermark: false
```

> ## Setting up your own archival fallback

Reviewer: ```suggestion
## Set up your own archival fallback
```

> It is also possible to set up the archival fallback to a different (personal) s3 bucket. For this a fullnode

Reviewer: You can also set up the archival fallback to use your own S3 bucket. To do so, you must configure your Full node to archive transactions and effects. You can enable this by adding the following config to your config file.

needs to be configured to archive transactions and effects. Enabling this can be done by adding the below config:

```yaml
state-archive-write-config:
  object-store-config:
    object-store: "S3"
    bucket: "<bucket_name>"
    aws-access-key-id: "<AWS_ACCESS_KEY_ID>"
    aws-secret-access-key: "<AWS_SECRET_ACCESS_KEY>"
    aws-region: "<aws_region>"
    object-store-connection-limit: 20
  concurrency: 5

>   # This is needed to be set as true on the node which is archiving the data

Reviewer: is it true or True?

on the node that archives the data

Author: it should be `true`

    ```

Your Full node starts on: `http://127.0.0.1:9000`.

# Pruning 

## Object Pruning

> Object pruning involves the removal of older object version from the db. New object versions are added to the db during transaction execution (making older versions ready for garbage collection) and can become a problem if not pruned, as it can slow down the performance of the db and take up too much disk space. Object pruning works in the background by tailing checkpoint stream where each checkpoint contains object versions which are now eligible for pruning.

Reviewer 2: easier to diff paragraphs with line breaks instead of spaces. They should just show up as spaces in the final formatting
```suggestion
Object pruning involves the removal of older object versions from the fullnode db.
New object versions that are added to the db during transaction execution (making older versions ready for garbage collection) can become a problem if not pruned. Many old object versions can slow down the performance of the db and take up large amounts of disk space.
Object pruning works in the background by tailing the checkpoint stream, identifying each checkpoint which contains object versions that are eligible for pruning.
```

  # This is recommended setting for validator nodes as older object versions aren't
  # needed for transaction execution.
  # When this is 1, pruner will only prune object versions from transaction checkpoints
  # older than the current epoch. In general, when this is N (where N >= 1), pruner 
  # will only prune object versions from checkpoints upto current - N epoch. 
  # This means it is possible to have multiple versions of an object to be present 
  # in the db. It is a recommended setting for fullnodes as they may have to serve 
  # rpc requests which require looking up objects by id and version (instead of just latest
  # version). However, if a fullnode is not serving such rpc requests it should also 
  # use aggressive pruning
  num-epochs-to-retain: 0
  # Advanced setting: Maximum number of checkpoints in one batch of pruning run. This is
  # mostly not needed to be changed and default is good for most use cases
  max-checkpoints-in-batch: 10
  # Advanced setting: Maximum number of transactions in one batch of pruning run. This is
  # mostly not needed to be changed and default is good for most use cases
  max-transactions-in-batch: 1000
```
## Transaction Pruning

> Transaction pruning is another type of pruning that removes old transactions and effects from the db. Checkpoints which are created periodically contain transaction and their effects. Transaction pruning works in the background by tailing checkpoint stream as well. A sui node can enable transaction pruning by adding `num_epochs_to_retain_for_checkpoints` to the `authority-store-pruning-config` config:

Reviewer 2: ```suggestion
Transaction pruning is another type of pruning that removes old transactions and effects from the db. Checkpoints which are created periodically contain transaction and their effects.
Transaction pruning works in the background by tailing checkpoint stream. 
A fullnodes and validators can enable transaction pruning by adding `num_epochs_to_retain_for_checkpoints` to their `authority-store-pruning-config` config:
```

  # When this is N (where N >= 2), pruner can prune transaction and effects from 
  # checkpoints upto current - N epoch. Transaction and effects from current and
  # previous epoch are never pruned. N = 2 is a recommended setting for validators
  # and fullnodes not serving rpc requests.
  num_epochs_to_retain_for_checkpoints: 2
  # Ensures that individual SST files periodically go through the compaction process.
  # This helps reclaim disk space and avoid fragmentation issues
  periodic-compaction-threshold-days: 1
```

## Archival Fallback

Once various full nodes start pruning historical transactions and effects, it may not be always possible for
their peers to catch up transaction and effects via state synchronization. And for this a node needs to fallback
downloading this data from a remote aws s3 bucket. This archive has all transaction and effects data since 
genesis and lags behind the latest network checkpoint by ten minutes. Enabling this is very highly recommended
for all nodes. A sui node can enable this automatic archive fallback by adding the following config:

```yaml
state-archive-read-config:
  - object-store-config:
      object-store: "S3"

>       # Use mysten-testnet-archives for testnet 

Reviewer 2: Can we add something on enabling a fullnode to create it's own archive? So ppl don't have to use mysten's buckets if they dont want to

Author: sure, will add

    ```

Your Full node starts on: `http://127.0.0.1:9000`.

> # Pruning 
> 
> ## Object Pruning

Reviewer: ```suggestion
## Object pruning
```

this would make 2 H1 in the topic, and is an empty heading

> Object pruning involves the removal of older object versions from the fullnode db. 

Reviewer: it's always Full node unless the specific database is named "fullnode".

Spell out database

> New object versions are added to the db during transaction execution (making older versions ready for garbage collection) and can become a 

Reviewer: passive voice
Sui adds new object versions to the database as part of transaction execution. This makes previous versions ready for garbage collection. However, without pruning, this can result in database performance degradation and requires  large amounts of storage space.
